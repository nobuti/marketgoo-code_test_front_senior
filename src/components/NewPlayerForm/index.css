.NewPlayerForm-field {
    margin-bottom: var(--size-4);
}

.NewPlayerForm-actions {
    display: flex;
    padding: 16px 0;
}

.NewPlayerForm-submit {
    margin-left: auto;
}
