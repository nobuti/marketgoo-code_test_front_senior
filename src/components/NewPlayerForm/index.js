import React, { useReducer } from 'react'
import { Field, Input, Button } from '@marketgoo/ola'
import PropTypes from 'prop-types'

import { required, moreThan, equalOrBiggerThan, validate, getError, anyError } from '../../utils/validations'

import './index.css'

const validations = {
  name: [required, moreThan(2)],
  team: [required, moreThan(2)],
  score: [equalOrBiggerThan(0)]
}

const Form = ({ onSubmit = () => {} }) => {
  const [state, dispatch] = useReducer((state, action) => {
    switch (action.type) {
      case 'change':
        return {
          ...state,
          touched: {
            ...state.touched,
            [action.field.name]: true
          },
          values: {
            ...state.values,
            [action.field.name]: action.field.value
          }
        }
      default:
        return state
    }
  }, {
    touched: {},
    values: {}
  })

  const onChange = (e) => {
    const { value, name } = e.target
    dispatch({
      type: 'change',
      field: {
        name,
        value
      }
    })
  }

  const onSubmitForm = e => {
    e.preventDefault()
    onSubmit(state.values)
  }

  const isThereAnyError = (errors, name) => {
    if (state.touched[name] == null) {
      return false
    }

    return getError(errors, name)
  }

  const errors = validate(state.values, validations)

  return <form onSubmit={onSubmitForm}>
    <div className="NewPlayerForm-field">
      <Field
        error={isThereAnyError(errors, 'name')}
        id="playerName"
        label="Name"
        className="NewPlayerForm-field"
      >
        <Input
          name="name"
          placeholder="Player name"
          type="text"
          onChange={onChange}
          data-testid="player-name"
        />
      </Field>
    </div>

    <div className="NewPlayerForm-field">
      <Field
        error={isThereAnyError(errors, 'team')}
        id="playerTeam"
        label="Team"
        className="NewPlayerForm-field"
      >
        <Input
          name="team"
          placeholder="Player team"
          type="text"
          onChange={onChange}
          data-testid="player-team"
        />
      </Field>
    </div>

    <div className="NewPlayerForm-field">
      <Field
        error={isThereAnyError(errors, 'score')}
        id="playerScore"
        label="Score"

      >
        <Input
          name="score"
          placeholder="Player score"
          type="number"
          onChange={onChange}
          data-testid="player-score"
        />
      </Field>
    </div>

    <div className="NewPlayerForm-actions">
      <Button variant="primary" disabled={anyError(errors)} type="submit" className="NewPlayerForm-submit" data-testid="player-submit">Add new player</Button>
    </div>
  </form>
}

Form.propTypes = {
  onSubmit: PropTypes.func
}

export default Form
