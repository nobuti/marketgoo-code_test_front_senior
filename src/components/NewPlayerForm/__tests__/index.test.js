import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'

import NewPlayerForm from '../'

describe('NewPlayerForm', () => {
  const onSubmit = jest.fn()

  test('renders properly', () => {
    const { container } = render(
      <NewPlayerForm />
    )

    expect(container).toMatchSnapshot()
  })

  test('submit should disable by default', () => {
    render(
      <NewPlayerForm />
    )

    expect(screen.queryByTestId('player-submit')).toBeDisabled()
  })

  test('submit should enable when data is valid', () => {
    render(
      <NewPlayerForm />
    )

    fireEvent.change(screen.queryByTestId('player-name'), {
      target: {
        value: 'Wadus'
      }
    })

    fireEvent.change(screen.queryByTestId('player-team'), {
      target: {
        value: 'Wadus'
      }
    })

    fireEvent.change(screen.queryByTestId('player-score'), {
      target: {
        value: '12'
      }
    })

    expect(screen.queryByTestId('player-submit')).toBeEnabled()
  })

  test('onSubmit callback should be called', () => {
    render(
      <NewPlayerForm onSubmit={onSubmit}/>
    )

    fireEvent.change(screen.queryByTestId('player-name'), {
      target: {
        value: 'Wadus'
      }
    })

    fireEvent.change(screen.queryByTestId('player-team'), {
      target: {
        value: 'Wadus'
      }
    })

    fireEvent.change(screen.queryByTestId('player-score'), {
      target: {
        value: '12'
      }
    })

    fireEvent.click(screen.queryByTestId('player-submit'))

    expect(onSubmit).toHaveBeenLastCalledWith({ name: 'Wadus', team: 'Wadus', score: '12' })
  })
})
