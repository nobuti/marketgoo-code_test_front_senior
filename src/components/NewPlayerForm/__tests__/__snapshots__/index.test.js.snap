// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPlayerForm renders properly 1`] = `
<div>
  <form>
    <div
      class="NewPlayerForm-field"
    >
      <div
        class="ola_field"
      >
        <label
          class="ola_field-label"
          for="playerName"
        >
          Name
        </label>
        <div
          class="ola_field-input"
        >
          <input
            class="ola_input"
            data-testid="player-name"
            id="playerName"
            name="name"
            placeholder="Player name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="NewPlayerForm-field"
    >
      <div
        class="ola_field"
      >
        <label
          class="ola_field-label"
          for="playerTeam"
        >
          Team
        </label>
        <div
          class="ola_field-input"
        >
          <input
            class="ola_input"
            data-testid="player-team"
            id="playerTeam"
            name="team"
            placeholder="Player team"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="NewPlayerForm-field"
    >
      <div
        class="ola_field"
      >
        <label
          class="ola_field-label"
          for="playerScore"
        >
          Score
        </label>
        <div
          class="ola_field-input"
        >
          <input
            class="ola_input"
            data-testid="player-score"
            id="playerScore"
            name="score"
            placeholder="Player score"
            type="number"
          />
        </div>
      </div>
    </div>
    <div
      class="NewPlayerForm-actions"
    >
      <button
        class="ola_button is-primary is-disabled NewPlayerForm-submit"
        data-testid="player-submit"
        disabled=""
        type="submit"
      >
        <span
          class="ola_button-text"
        >
          Add new player
        </span>
      </button>
    </div>
  </form>
</div>
`;
